packages/flutter/lib/src/widgets/raster_widget.dart

```diff
@@ -0,0 +1,258 @@
+  /// delegate class (which amounts to the same thing, because the latter is
+  /// implemented in terms of the former).
+  ///
+  /// If the new instance represents different information than the old
+  /// instance, then the method should return true, otherwise it should return
+  /// false.
+  ///
+  /// If the method returns false, then the [paint] call might be optimized
+  /// away.
+  ///
+  /// It's possible that the [paint] method will get called even if
+  /// [shouldRepaint] returns false (e.g. if an ancestor or descendant needed to
+  /// be repainted). It's also possible that the [paint] method will get called
+  /// without [shouldRepaint] being called at all (e.g. if the box changes
+  /// size).
+  ///
+  /// The `oldDelegate` argument will never be null.
+  bool shouldRepaint(covariant RasterWidgetDelegate oldDelegate);
+}
+
+/// A render object that draws its child as a [ui.Image].
+class RenderRasterWidget extends RenderProxyBox {
+  /// Create a new [RenderRasterWidget].
+  RenderRasterWidget({
+    required RasterWidgetDelegate delegate,
+    required double devicePixelRatio,
+    required ValueNotifier<bool> rasterize,
+  }) : _delegate = delegate,
+       _devicePixelRatio = devicePixelRatio,
+       _rasterize = rasterize {
+    _currentlyRepaintBoundary = rasterize.value;
+  }
+
+  /// The device pixel ratio used to create the child image.
+  double get devicePixelRatio => _devicePixelRatio;
+  double _devicePixelRatio;
+  set devicePixelRatio(double value) {
+    if (value == devicePixelRatio) {
+      return;
+    }
+    _devicePixelRatio = value;
+    markNeedsPaint();
+  }
+
+  /// Whether a rasterized version of this render objects child is drawn in
+  /// place of the child.
+  ValueNotifier<bool> get rasterize => _rasterize;
+  ValueNotifier<bool> _rasterize;
+  set rasterize(ValueNotifier<bool> value) {
+    if (value == rasterize) {
+      return;
+    }
+    rasterize.removeListener(_onRasterValueChanged);
+    final bool oldValue = rasterize.value;
+    _rasterize = value;
+    rasterize.addListener(_onRasterValueChanged);
+    if (oldValue != rasterize.value) {
+      _onRasterValueChanged();
+    }
+  }
+
+  /// The delegate used to draw the image representing the child.
+  RasterWidgetDelegate get delegate => _delegate;
+  RasterWidgetDelegate _delegate;
+  set delegate(RasterWidgetDelegate value) {
+    if (value == delegate) {
+      return;
+    }
+    delegate.removeListener(markNeedsPaint);
+    final RasterWidgetDelegate oldDelegate = _delegate;
+    _delegate = value;
+    delegate.addListener(markNeedsPaint);
+    if (delegate.shouldRepaint(oldDelegate)) {
+      markNeedsPaint();
+    }
+  }
+
+  ui.Image? _childRaster;
+  bool _currentlyRepaintBoundary = false;
+
+  @override
+  void attach(covariant PipelineOwner owner) {
+    delegate.addListener(markNeedsPaint);
+    rasterize.addListener(_onRasterValueChanged);
```

Comment:
These get added twice if deelgate and rasterizer are set while the RO is detached.

Reply:
I always miss this one